readcomb/phase_change_filter.py

```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
+        return snp_lst
+    
     for snp in snps:
-        start = snp.start - record.reference_start # snp.start is 0-based
-        
+        # Using SNP.start and record.reference_start since they are both 0 based
+        # SNP.start grabs vcf positions in 0 index while vcfs are 1 indexed
+
+        start = snp.start - record.reference_start
+
+        # extra calculations to realign start if there is an insertion or deletion
         current_tuple = 0
         current_base = 0
-        
+
         while current_base < start and current_tuple < len(cigar_tuples):
             if cigar_tuples[current_tuple][0] == 1:
-                # shift the start over by the amount of insertion to compensate for it
+                # shift the start to the right by the amount of insertion to compensate for it
                 start += cigar_tuples[current_tuple][1]
             
+            elif cigar_tuples[current_tuple][0] == 2:
+                # shift the start to the left by the amount of deletion to compensate for it
+                start -= cigar_tuples[current_tuple][1]
+
             current_base += cigar_tuples[current_tuple][1]
             current_tuple += 1
-    
+
+
         # indexing for VCF seems to be a bit weird and will sometimes be -1
+        # this has been fixed but left check in anyways
         if start < 0:
             raise Exception('VCF indexing is off. Check SNP at {}'.format(snp))
-        
+
         strand1 = snp.gt_bases[0][0]
         strand2 = snp.gt_bases[1][0]
-        
+
         if start >= len(segment):
             break
-        
+
         if segment[start] == strand1:
-            snp_lst[start] = '1'
+            snp_lst.append('1')
+
         elif segment[start] == strand2:
-            snp_lst[start] = '2'
+            snp_lst.append('2')
+
         else:
-            snp_lst[start] = 'N'
+            snp_lst.append('N')
+
+    return snp_lst
+
+def matepairs_recomb():
+    '''
+    Given a bam file object, return a human-readable file with aligned reference, quality string, 
+        bam sequence, and SNPs    
+
+    If mode='no_match', create a file with only no_match sequences.
+    If mode='all', create a file with all sequences
+    If mode='phase_change', create a file with only sequences with phase changes
+    
+    Makes assumptions on the following file locations:
+        VCF: vcf_files_path + 'parental_filtered.vcf.gz'
+        Reference: reference_files_path + 'chlamy.5.3.w_organelles_mtMinus.fasta'
+    '''
+
+    bam, vcf, mode, output_filename = args()
+
+    bam_file_obj = pysam.AlignmentFile(bam, 'r')
+    
+    f_obj = pysam.AlignmentFile(output_filename + '.sam', 'wh', template=bam_file_obj)
+    
+    #counters
+    no_match_counter = 0
+    phase_change_counter = 0
+    phase_change_mate_pair_counter = 0
+    seq_with_snps_counter = 0
+    all_seq_counter = 0
+    
+    # get mate pairs
+    pairs = cache_pairs(bam_file_obj)
+    
+    for query_name in pairs:
```

Comment:
Here's what I was thinking for this part:

```python
for query_name in pairs:
        
        all_seq_counter += 1
        record_counter = 0
        snp_lst = []
        
        for record in pairs[query_name]:
            if record:
                record_counter += 1
            
                # analyze cigar string
                segment = cigar(record)
                            
                snps = check_snps(vcf, record.reference_name, 
                                record.reference_start,
                                record.reference_start + record.query_alignment_length)
                
                snp_lst += phase_detection(snps, segment, record)

                if 'N' in snp_lst:
                    no_match_counter += 1

                    if 'no_match' in mode:                        
                        f_obj.write(record)

        #phase_change_counter update
        if '1' in snp_lst and '2' in snp_lst:
            phase_change_counter += 1

            if 'phase_change' in mode:                        
                if record_counter == 1:
                    f_obj.write(pairs[query_name][0])
                elif record_counter == 2:
                    for record in pairs[query_name]:
                        f_obj.write(record)

```

It makes the code a lot simpler by consolidating things into a single snp list, and then just doing a few checks for whether the second record exists where needed. We can chat more about it on Thursday

Reply:
This is actually really clean, I guess I really tunnel visioned when I rewrote this. I'll change the code and test it